### DISCUSSION TITLE
Information about Samples data in mqtt

### DISCUSSION THREAD

**Discussion by emvarunp** [mtconnect/cppagent#470] [General]
Hi @wsobel,

I'm looking for any information or documentation on how the samples data will be sent in mqtt sink. In the REST sink we see that samples will give a history of the dataitems. But in mqtt since there is no way to send a req to the agent, how will the agent send the samples data.

Thank you for any guidance or information,
V

**Comment by wsobel** [comment]
Are you using the first or second implementation? the mqtt2 sink is probably a better bet right now. It may answer you questions

**Reply by emvarunp** [reply]
im using mqtt2.
does agent publish history of dataitem like it does in the REST sink. or does it publish it in any other way.

Thanks for the reply @wsobel.

**Reply by wsobel** [reply]
We do as much as we can with MQTT. MQTT does not support history the way the agent does. That is their implementation decision and I can't change it. 

I do support key frame style reporting in the implementation. If you get the current topic, that gives the snapshot for all the data items at that point in time. The sample topic gives the time series delta at a faster interval. You can adjust the publishing rates to your needs. 

This gives a limited amount of history, but it's the best I could do without store-and-forward in MQTT.

**Reply by emvarunp** [reply]
> The sample topic gives the time series delta at a faster interval.

This is what i was looking for.
Thanks for the info @wsobel,
V
